README.md
# <a name="title"></a> Kitchen::Ec2: A Test Kitchen Driver for Amazon EC2

A [Test Kitchen][kitchenci] Driver for Amazon EC2.

This driver uses the [fog gem][fog_gem] to provision and destroy EC2
instances. Use Amazon's cloud for your Linux and Windows infrastructure testing!

## <a name="requirements"></a> Requirements

There are **no** external system requirements for this driver. However you
will need access to an [AWS][aws_site] account.

## <a name="installation"></a> Installation and Setup

Please read the [Driver usage][driver_usage] page for more details.

## <a name="default-config"></a> Default Configuration

This driver can determine AMI and username login for a select number of
platforms in each region.

For Windows instances the generated Administrator password is fetched automatically from Amazon EC2 with the same private key as we use for SSH logins to Linux. Currently, the following platform names are
supported:

```ruby
---
platforms:
  - name: ubuntu-10.04
  - name: ubuntu-12.04
  - name: ubuntu-12.10
  - name: ubuntu-13.04
  - name: centos-6.4
  - name: debian-7.1.0
  - name: windows-2008r2
  - name: windows-2012r2

```


This will effectively generate a configuration similar to:

```ruby
---
platforms:
  - name: ubuntu-10.04
    driver:
      image_id: ami-1ab3ce73
      username: ubuntu
  - name: ubuntu-12.04
    driver:
      image_id: ami-2f115c46
      username: ubuntu
  # ...
  - name: centos-6.4
    driver:
      image_id: ami-bf5021d6
      username: root
  # ...
```

For specific default values, please consult [amis.json][amis_json].

## <a name="config"></a> Configuration

### <a name="config-az"></a> availability\_zone

**Required** The AWS [availability zone][region_docs] to use.

The default is `"us-east-1b"`.

### <a name="config-aws-access-key-id"></a> aws\_access\_key\_id

**Required** The AWS [access key id][credentials_docs] to use.

The default will be read from the `AWS_ACCESS_KEY` environment variable if set,
or `nil` otherwise.

### <a name="config-aws-secret-access-key"></a> aws\_secret\_access\_key

**Required** The AWS [secret access key][credentials_docs] to use.

The default will be read from the `AWS_SECRET_KEY` environment variable if set,
or `nil` otherwise.

### <a name="config-aws-ssh-key-id"></a> aws\_ssh\_key\_id

**Required** The EC2 [SSH key id][key_id_docs] to use.

The default will be read from the `AWS_SSH_KEY_ID` environment variable if set,
or `nil` otherwise.

### <a name="config-aws-session-token"></a> aws\_session\_token

The AWS [session token][credentials_docs] to use.

The default will be read from the `AWS_SESSION_TOKEN` environment variable if set,
or `nil` otherwise.

### endpoint

The API endpoint for executing EC2 commands.

The default will be computed from the AWS region name for the instance.

### <a name="config-flavor-id"></a> flavor\_id

The EC2 [instance type][instance_docs] (also known as size) to use.

The default is `"m1.small"`.

### <a name="config-ebs-optimized"></a> ebs\_optimized

Option to launch EC2 instance with optimized EBS volume. See
[Amazon EC2 Instance Types](http://aws.amazon.com/ec2/instance-types/) to find
out more about instance types that can be launched as EBS-optimized instances.

The default is `false`.

### <a name="config-security-group-ids"></a> security_group_ids

An Array of EC2 [security groups][group_docs] which will be applied to the
instance.

The default is `["default"]`.

### <a name="config-image-id"></a> image\_id

**Required** The EC2 [AMI id][ami_docs] to use.

The default will be determined by the `aws_region` chosen and the Platform
name, if a default exists (see [amis.json][ami_json]). If a default cannot be
computed, then the default is `nil`.

### <a name="config-port"></a> port

The SSH port number to be used when communicating with the instance.

The default is `22`.

### <a name="interface"></a> interface

The place from which to derive the hostname for communicating with the instance.  May be `dns`, `public` or `private`.  If this is unset, the driver will derive the hostname by failing back in the following order:

1. DNS Name
2. Public IP Address
3. Private IP Address

The default is unset.


### <a name="config-region"></a> region

**Required** The AWS [region][region_docs] to use.

The default is `"us-east-1"`.

### <a name="config-ssh-key"></a> ssh\_key

Path to the private SSH key used to connect to the instance.

The default is unset, or `nil`.

### <a name="config-subnet-id"></a> subnet\_id

The EC2 [subnet][subnet_docs] to use.

The default is unset, or `nil`.

### <a name="config-tags"></a> tags

The Hash of EC tag name/value pairs which will be applied to the instance.

The default is `{ "created-by" => "test-kitchen" }`.

### <a name="config-username"></a> username

The SSH or WinRM username that will be used to communicate with the instance.

The default will be determined by the Platform name, if a default exists (see   [amis.json][amis_json]). If a default cannot be computed, then the default is `"root"`.

For Windows AMIs if the username is configured other then Administrator,
kitchen-ec2 will create that user during the server creation with the specified
static password.

### <a name="config-password"></a> password

WinRM static password  to be used for connecting to a Windows AMI with the
configured static `username` other than Administrator.

### <a name="config-iam-profile-name"></a> iam\_profile\_name

The EC2 IAM profile name to use.

The default is `nil`.

## <a name="example"></a> Example

The following could be used in a `.kitchen.yml` or in a `.kitchen.local.yml`
to override default configuration.

```yaml
---
driver:
  name: ec2
  aws_access_key_id: KAS...
  aws_secret_access_key: 3UK...
  aws_ssh_key_id: id_rsa-aws
  ssh_key: /path/to/id_rsa-aws
  security_group_ids: ["sg-1a2b3c4d"]
  region: us-east-1
  availability_zone: us-east-1b
  require_chef_omnibus: true
  subnet_id: subnet-6d6...
  iam_profile_name: chef-client
  tags:
    Name: 'test-kitchen instance'
    Environment: 'dev'

provisioner:
  name: chef_solo

# Default transport is SSH, which should be overridden for Windows to winrm
transport:
  name: ssh

platforms:
  - name: ubuntu-12.04
    driver:
      image_id: ami-fd20ad94
      username: ubuntu
  - name: centos-6.3
    driver:
      image_id: ami-ef5ff086
      username: ec2-user
  - name: windows-2012R2
    transport:
      name: winrm
  - name: windows-2008R2
    transport:
      name: winrm
      max_threads: 4

suites:
# ...
```

Both `.kitchen.yml` and `.kitchen.local.yml` files are pre-processed through ERB which can help to factor out secrets and credentials. For example:

```yaml
---
driver:
  name: ec2
  aws_access_key_id: <%= ENV['AWS_ACCESS_KEY'] %>
  aws_secret_access_key: <%= ENV['AWS_SECRET_KEY'] %>
  aws_ssh_key_id: <%= ENV['AWS_SSH_KEY_ID'] %>
  ssh_key: <%= File.expand_path('~/.ssh/id_rsa') %>
  security_group_ids: ["sg-1a2b3c4d"]
  region: us-east-1
  availability_zone: us-east-1b
  require_chef_omnibus: true

platforms:
  - name: ubuntu-12.04
    driver:
      image_id: ami-fd20ad94
      username: ubuntu
  - name: centos-6.3
    driver:
      image_id: ami-ef5ff086
      username: ec2-user

suites:
# ...
```

## <a name="development"></a> Development

* Source hosted at [GitHub][repo]
* Report issues/questions/feature requests on [GitHub Issues][issues]

Pull requests are very welcome! Make sure your patches are well tested.
Ideally create a topic branch for every separate change you make. For
example:

1. Fork the repo
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## <a name="authors"></a> Authors
Created and maintained by [Fletcher Nichol][author] (<fnichol@nichol.ca>)
Windows AMI support added by [Salim Afiune][author] (<afiune@getchef.com>) and [Akos Murati][author] (<akos@murati.hu>)

## <a name="license"></a> License

Apache 2.0 (see [LICENSE][license])


[author]:           https://github.com/fnichol
[issues]:           https://github.com/muratiakos/kitchen-ec2/issues
[license]:          https://github.com/muratiakos/kitchen-ec2/blob/master/LICENSE
[repo]:             https://github.com/muratiakos/kitchen-ec2
[driver_usage]:     http://docs.kitchen-ci.org/drivers/usage
[chef_omnibus_dl]:  http://www.getchef.com/chef/install/

[amis_json]:        https://github.com/muratiakos/kitchen-ec2/blob/master/data/amis.json
[ami_docs]:         http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ComponentsAMIs.html
[aws_site]:         http://aws.amazon.com/
[credentials_docs]: http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/SettingUp_CommandLine.html#using-credentials-access-key
[fog_gem]:          http://fog.io/
[group_docs]:       http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-network-security.html
[instance_docs]:    http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-types.html
[key_id_docs]:      http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/verifying-your-key-pair.html
[kitchenci]:        http://kitchen.ci/
[region_docs]:      http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html
[subnet_docs]:      http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-subnet.html


CHANGELOG.md
## 0.9.3 / 2015-02-13
### Bug fixes
* Removing default availability-zone to let Amazon defaults to be applied for new instances

## 0.9.2 / 2015-01-27
### New features
* Adding support for AWS Private Key Pair from TeamCity through AWS_PRIVATE_KEY Environment variable

## 0.9.1 / 2015-01-20
### New features
* Supports Amazon generated Windows Password
* Improved WinRm provisioning through user_data


## 0.8.0 / 2014-02-11

### Bug fixes

* Pull request [#29][], pull request [#28][]: Relax Test Kitchen dependency to allow versions 1.0 and up. ([@coderanger][], [@someara][])

## New features

* Pull request [#29][]: Add `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` fallback environment variable support. ([@coderanger][])
* Pull request [#27][], pull request [#31][]: Make EC2 endpoint configurable for services implementing the AWS APIs. ([@bozinsky][], [@spheromak][])
* Pull request [#34][]: Support AWS session tokens for use with IAM roles. ([@coderanger][])
* Pull request [#22][]: Add ebs_optimized attribute. ([@tiwilliam][])

### Improvements

* Pull request [#20][], pull request [#21][], issue [#14][]: Add configurability around whether to use DNS name, public or private IP addresses when connecting to instances. ([@matheeeny][], [@Atalanta][], [@fnichol][])
* Pull request [#15][]: In VPC groups must be specified by setting :security_group_ids rather than :groups. ([@eherot][])
* Pull request [#23][]: README updates for badges. ([@sethvargo][], [@arangamani][])


## 0.7.0 / 2013-08-29

### Bug fixes

* Pull request [#13][]: #wait_for_ssh takes 2 arguments. ([@dysinger][])

### Improvements

* Support computed defaults for a select list of pre-determined platforms (see readme for quick example). ([@fnichol][])~


## 0.6.0 / 2013-07-23

### Bug fixes

* Pull request [#8][]: Use private ip if the public ip is `nil`. ([@dissonanz][])

### Improvements

* Pull request [#9][]: Match access and secret key env vars in example kitchen config with CLI tools' env vars. ([@juliandunn][])


## 0.5.1 / 2013-05-23

### New features

* Pull request [#7][]: Add subnet\_id option for use with VPC. ([@dissonanz][])


## 0.5.0 / 2013-05-23

### New features

* Pull request [#5][]: Add the ability to give ec2 instances tags. ([@halcyonCorsair][])
* Add required_config attributes for driver. ([@fnichol][])

### Improvements

* Write README. ([@fnichol][])
* Pull request [#2][]: Extra EC2 server creation debugging. ([@mattray][])
* Remove default_config :port in favor of SSHBase default (also 22). ([@fnichol][])

<!--- The following link definition list is generated by PimpMyChangelog --->
[#2]: https://github.com/opscode/kitchen-ec2/issues/2
[#5]: https://github.com/opscode/kitchen-ec2/issues/5
[#7]: https://github.com/opscode/kitchen-ec2/issues/7
[#8]: https://github.com/opscode/kitchen-ec2/issues/8
[#9]: https://github.com/opscode/kitchen-ec2/issues/9
[#13]: https://github.com/opscode/kitchen-ec2/issues/13
[#14]: https://github.com/opscode/kitchen-ec2/issues/14
[#15]: https://github.com/opscode/kitchen-ec2/issues/15
[#20]: https://github.com/opscode/kitchen-ec2/issues/20
[#21]: https://github.com/opscode/kitchen-ec2/issues/21
[#22]: https://github.com/opscode/kitchen-ec2/issues/22
[#23]: https://github.com/opscode/kitchen-ec2/issues/23
[#27]: https://github.com/opscode/kitchen-ec2/issues/27
[#28]: https://github.com/opscode/kitchen-ec2/issues/28
[#29]: https://github.com/opscode/kitchen-ec2/issues/29
[#31]: https://github.com/opscode/kitchen-ec2/issues/31
[#34]: https://github.com/opscode/kitchen-ec2/issues/34
[@Atalanta]: https://github.com/Atalanta
[@arangamani]: https://github.com/arangamani
[@bozinsky]: https://github.com/bozinsky
[@coderanger]: https://github.com/coderanger
[@dissonanz]: https://github.com/dissonanz
[@dysinger]: https://github.com/dysinger
[@eherot]: https://github.com/eherot
[@fnichol]: https://github.com/fnichol
[@halcyonCorsair]: https://github.com/halcyonCorsair
[@juliandunn]: https://github.com/juliandunn
[@matheeeny]: https://github.com/matheeeny
[@mattray]: https://github.com/mattray
[@sethvargo]: https://github.com/sethvargo
[@someara]: https://github.com/someara
[@spheromak]: https://github.com/spheromak
[@tiwilliam]: https://github.com/tiwilliam


lib/kitchen/driver/ec2_windows.rb
# -*- encoding: utf-8 -*-
#
# Author:: Salim Afiune (<salim@afiunemaya.com.mx>)
#
# Copyright (C) 2014, Salim Afiune
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'benchmark'
require 'json'
require 'fog'
require 'kitchen'

module Kitchen

  module Driver

    # Amazon EC2 driver for Test Kitchen using WinRM protocol.
    #
    # @author Salim Afiune <salim@afiunemaya.com.mx>
    class Ec2Windows < Kitchen::Driver::WinRMBase

      default_config :region,             'us-east-1'
      default_config :availability_zone,  'us-east-1b'
      default_config :flavor_id,          'm1.small'
      default_config :ebs_optimized,      false
      default_config :security_group_ids, ['default']
      default_config :tags,               { 'created-by' => 'test-kitchen' }
      default_config :iam_profile_name,   nil
      default_config :aws_access_key_id do |driver|
        ENV['AWS_ACCESS_KEY'] || ENV['AWS_ACCESS_KEY_ID']
      end
      default_config :aws_secret_access_key do |driver|
        ENV['AWS_SECRET_KEY'] || ENV['AWS_SECRET_ACCESS_KEY']
      end
      default_config :aws_session_token do |driver|
        ENV['AWS_SESSION_TOKEN'] || ENV['AWS_TOKEN']
      end

      default_config :image_id do |driver|
        driver.default_ami
      end
   
      default_config :username do |driver|
        driver.default_username
      end
   
      default_config :password do |driver|
        driver.default_password
      end
   
      default_config :endpoint do |driver|
        "https://ec2.#{driver[:region]}.amazonaws.com/"
      end

      default_config :interface, nil

      required_config :aws_access_key_id
      required_config :aws_secret_access_key
      required_config :image_id

      def create(state)
        return if state[:server_id]
        server = create_server
        state[:server_id] = server.id
        state[:password]  = config[:password]

        info("EC2 instance <#{state[:server_id]}> created.")
        server.wait_for { print '.'; ready? }
        print ' '
        info '[Server] Ready.'
        state[:hostname] = hostname(server)
        wait_for_winrm(state[:hostname], config[:username], state)
        info '[WinRM] Ready.'
        debug("ec2:create '#{state[:hostname]}'")
      rescue Fog::Errors::Error, Excon::Errors::Error => ex
        raise ActionFailed, ex.message
      end

      def destroy(state)
        return if state[:server_id].nil?

        server = connection.servers.get(state[:server_id])
        server.destroy unless server.nil?
        info("EC2 instance <#{state[:server_id]}> destroyed.")
        state.delete(:server_id)
        state.delete(:hostname)
      end

      def default_ami
        region = amis['regions'][config[:region]]
        region && region[instance.platform.name]
      end

      def default_username
        amis['usernames'][instance.platform.name] || 'Administrator'
      end

      def default_password
        'chef'
      end

      private

      def connection
        Fog::Compute.new(
          :provider               => :aws,
          :aws_access_key_id      => config[:aws_access_key_id],
          :aws_secret_access_key  => config[:aws_secret_access_key],
          :aws_session_token      => config[:aws_session_token],
          :region                 => config[:region],
          :endpoint               => config[:endpoint],
        )
      end

      def create_server
        debug_server_config

        connection.servers.create(
          :availability_zone         => config[:availability_zone],
          :security_group_ids        => config[:security_group_ids],
          :tags                      => config[:tags],
          :flavor_id                 => config[:flavor_id],
          :ebs_optimized             => config[:ebs_optimized],
          :image_id                  => config[:image_id],
          :subnet_id                 => config[:subnet_id],
          :iam_instance_profile_name => config[:iam_profile_name],
          :user_data                 => winrm_config
        )
      end

      def debug_server_config
        debug("ec2:region '#{config[:region]}'")
        debug("ec2:availability_zone '#{config[:availability_zone]}'")
        debug("ec2:flavor_id '#{config[:flavor_id]}'")
        debug("ec2:ebs_optimized '#{config[:ebs_optimized]}'")
        debug("ec2:image_id '#{config[:image_id]}'")
        debug("ec2:security_group_ids '#{config[:security_group_ids]}'")
        debug("ec2:tags '#{config[:tags]}'")
        debug("ec2:subnet_id '#{config[:subnet_id]}'")
        debug("ec2:iam_profile_name '#{config[:iam_profile_name]}'")
      end

      def amis
        @amis ||= begin
          json_file = File.join(File.dirname(__FILE__),
            %w{.. .. .. data amis.json})
          JSON.load(IO.read(json_file))
        end
      end

      def interface_types
        {
          'dns' => 'dns_name',
          'public' => 'public_ip_address',
          'private' => 'private_ip_address'
        }
      end

      def winrm_config
        <<-EOH.gsub(/^ {10}/, '')
        <powershell>
          $username="#{config[:username]}"
          $password="#{config[:password]}"

          $seccfg = [IO.Path]::GetTempFileName()
          secedit /export /cfg $seccfg
          (Get-Content $seccfg) | Foreach-Object {$_ -replace "PasswordComplexity\\s*=\\s*1", "PasswordComplexity=0"} | Set-Content $seccfg
          secedit /configure /db $env:windir\\security\\new.sdb /cfg $seccfg /areas SECURITYPOLICY
          del $seccfg
           
          net user /add $username $password;
          net localgroup Administrators /add $username;

          try { winrm quickconfig -q }
          catch {write-host "winrm quickconfig failed"}
          winrm set winrm/config/winrs '@{MaxMemoryPerShellMB="512"}'
          winrm set winrm/config '@{MaxTimeoutms="1800000"}'
          winrm set winrm/config/service '@{AllowUnencrypted="true"}'
          winrm set winrm/config/service/auth '@{Basic="true"}'
          netsh advfirewall firewall set rule name="Windows Remote Management (HTTP-In)" profile=public protocol=tcp localport=5985 remoteip=localsubnet new remoteip=any

        </powershell>
        EOH
      end

      def hostname(server)
        if config[:interface]
          method = interface_types.fetch(config[:interface]) do
            raise Kitchen::UserError, 'Invalid interface'
          end
          server.send(method)
        else
          server.dns_name || server.public_ip_address || server.private_ip_address
        end
      end
    end
  end
end


lib/kitchen/driver/ec2.rb
# -*- encoding: utf-8 -*-
#
# Author:: Fletcher Nichol (<fnichol@nichol.ca>)
#
# Copyright (C) 2012, Fletcher Nichol
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'benchmark'
require 'json'
require 'fog'
require 'kitchen'

module Kitchen

  module Driver

    # Amazon EC2 driver for Test Kitchen with native
    # AWS generated Windows password support
    #
    # @author Fletcher Nichol <fnichol@nichol.ca>
    class Ec2 < Kitchen::Driver::Base

      default_config :region,             'us-east-1'
      default_config :availability_zone,  nil
      default_config :flavor_id,          'm1.small'
      default_config :ebs_optimized,      false
      default_config :security_group_ids, ['default']
      default_config :tags,               { 'created-by' => 'test-kitchen' }
      default_config :iam_profile_name,   nil
      default_config :aws_access_key_id do |driver|
        ENV['AWS_ACCESS_KEY'] || ENV['AWS_ACCESS_KEY_ID']
      end
      default_config :aws_secret_access_key do |driver|
        ENV['AWS_SECRET_KEY'] || ENV['AWS_SECRET_ACCESS_KEY']
      end
      default_config :aws_session_token do |driver|
        ENV['AWS_SESSION_TOKEN'] || ENV['AWS_TOKEN']
      end
      default_config :aws_ssh_key_id do |driver|
        ENV['AWS_SSH_KEY_ID']
      end
      default_config :image_id do |driver|
        driver.default_ami
      end
      default_config :username do |driver|
        driver.default_username
      end
      default_config :password, ''

      default_config :endpoint do |driver|
        "https://ec2.#{driver[:region]}.amazonaws.com/"
      end

      default_config :interface, nil

      required_config :aws_access_key_id
      required_config :aws_secret_access_key
      required_config :aws_ssh_key_id
      required_config :image_id

      #Creating a new instance
      def create(state)
        return if state[:server_id]
        state[:port] = default_port
        state[:password] = config[:password] if config[:password]
        state[:username] = config[:username] if config[:username]

        server = create_server
        state[:server_id] = server.id

        info("EC2 instance <#{state[:server_id]}> created.")

        #Server preparation
        debug('Waiting for Server')
        server.wait_for { print '.'; ready? }
        print '(Server Ready)'
        state[:hostname] = hostname(server)

        #Windows preparartion
        if transport.name.casecmp('winrm') == 0
          debug("Waiting for Windows")
          $stdout.sync = true
          while !windows_ready?(state)
            putc '.'
            sleep(10)
          end
          print '(Windows Ready)'

          #If we want to use the EC2 generated Admin password
          if config[:username].casecmp('Administrator') != 0
            print("Using static credentials")
          else
            print('Fetching EC2 generated credentials')
            config[:password] = windows_password(state)
            state[:password] = config[:password]
          end
        end
        debug("Credentials: #{state[:username]} #{state[:password]}")

        debug("Waiting for transport")
        transport.connection(state) do |c|
          c.wait_for_connection
        end
        print '(Transport Ready)'


        debug("ec2:create '#{state[:hostname]}'")
      rescue Fog::Errors::Error, Excon::Errors::Error => ex
        raise ActionFailed, ex.message
      end

      #Terminating the test instance
      def destroy(state)
        return if state[:server_id].nil?

        server = connection.servers.get(state[:server_id])
        server.destroy unless server.nil?
        info("EC2 instance <#{state[:server_id]}> destroyed.")
        state.delete(:server_id)
        state.delete(:hostname)
      end

      # Helper method to map a regial AMI for the OS
      def default_ami
        region = amis['regions'][config[:region]]
        region && region[instance.platform.name.downcase]
      end

      # Helper method to choose the admin user for the OS
      def default_username
        amis['usernames'][instance.platform.name.downcase] || 'root'
      end

      private

      # Fog AWS helper method for creating connection
      def connection
        Fog::Compute.new(
        :provider               => :aws,
        :aws_access_key_id      => config[:aws_access_key_id],
        :aws_secret_access_key  => config[:aws_secret_access_key],
        :aws_session_token      => config[:aws_session_token],
        :region                 => config[:region],
        :endpoint               => config[:endpoint],
        )
      end

      # Fog AWS helper for creating the instance
      def create_server
        debug_server_config

        debug('Creating EC2 Instance..')
        connection.servers.create(
        :availability_zone         => config[:availability_zone],
        :security_group_ids        => config[:security_group_ids],
        :tags                      => config[:tags],
        :flavor_id                 => config[:flavor_id],
        :ebs_optimized             => config[:ebs_optimized],
        :image_id                  => config[:image_id],
        :key_name                  => config[:aws_ssh_key_id],
        :subnet_id                 => config[:subnet_id],
        :iam_instance_profile_name => config[:iam_profile_name],
        :user_data                 => prepared_user_data
        )
      end

      # Method for preparing user_data for enabling PS Remoting if the selected
      # transport method is WinRM

      def prepared_user_data
        if transport.name.casecmp('winrm') == 0
          debug("Injecting WinRM config to EC2 user_data")

          #Preparing custom static admin user if we defined something other than Administrator
          customAdminScript = ''
          if config[:username].casecmp('Administrator') != 0
            debug('Injecting custom Local Administrator:')
            debug("username '#{config[:username]}'")
            debug("password '#{config[:password]}'")

            customAdminScript = <<-EOH.gsub(/^ {10}/, '')
            "Disabling Complex Passwords" >> $logfile
            $seccfg = [IO.Path]::GetTempFileName()
            & secedit.exe /export /cfg $seccfg >> $logfile
            (Get-Content $seccfg) | Foreach-Object {$_ -replace "PasswordComplexity\\s*=\\s*1", "PasswordComplexity = 0"} | Set-Content $seccfg
            & secedit.exe /configure /db $env:windir\\security\\new.sdb /cfg $seccfg /areas SECURITYPOLICY >> $logfile
            & cp $seccfg "c:\\"
            & del $seccfg

            $username="#{config[:username]}"
            $password="#{config[:password]}"

            "Creating static user: $username" >> $logfile
            & net.exe user /y /add $username $password >> $logfile

            "Adding $username to Administrators" >> $logfile
            & net.exe localgroup Administrators /add $username >> $logfile

            EOH
          end

          # Returning the fully constructed PowerShell script to user_data
          return <<-EOH.gsub(/^ {10}/, '')
          <powershell>
          $logfile="C:\\Program Files\\Amazon\\Ec2ConfigService\\Logs\\kitchen-ec2.log"

          #PS Remoting and & winrm.cmd basic config
          Enable-PSRemoting -Force -SkipNetworkProfileCheck
          & winrm.cmd quickconfig -q >> $logfile
          & winrm.cmd set winrm/config '@{MaxTimeoutms="1800000"}' >> $logfile
          & winrm.cmd set winrm/config/winrs '@{MaxMemoryPerShellMB="512"}' >> $logfile
          & winrm.cmd set winrm/config/winrs '@{MaxShellsPerUser="50"}' >> $logfile

          #Client settings
          & winrm.cmd set winrm/config/client/auth '@{Basic="true"}' >> $logfile

          #Server settings
          & winrm.cmd set winrm/config/service/auth '@{Basic="true"}' >> $logfile
          & winrm.cmd set winrm/config/service '@{AllowUnencrypted="true"}' >> $logfile
          & winrm.cmd set winrm/config/winrs '@{MaxMemoryPerShellMB="1024"}' >> $logfile

          #Firewall Config
          & netsh.exe advfirewall set publicprofile state off  >> $logfile
          & netsh advfirewall firewall set rule name="Windows Remote Management (HTTP-In)" profile=public protocol=tcp localport=5985 remoteip=localsubnet new remoteip=any  >> $logfile

          #{customAdminScript}

          #{config[:user_data]}

          </powershell>
          EOH
        else
          return config[:user_data] if !config[:user_data].nil?
        end
        return ''
      end

      # Helper method to check whether Amazon reported a server Ready
      def windows_ready? state
        log = connection.get_console_output(state[:server_id]).data[:body]["output"]
        if !log.nil?
          debug("Console output: --- \n")
          debug(log)
        end
        !log.nil? and log =~ /Windows is Ready to use/
      end

      # Helper method to fetch and decrypt Windows password from EC2
      def windows_password state
        enc = connection.get_password_data(state[:server_id]).data[:body]["passwordData"].strip!
        enc = Base64.decode64(enc)
        rsa = OpenSSL::PKey::RSA.new aws_private_key
        rsa.private_decrypt(enc) if !enc.nil? and enc != ''
      rescue NoMethodError
        debug('Unable to fetch encrypted password')
        return ''
      rescue TypeError
        debug('Unable to decrypt password with AWS_PRIVATE_KEY')
        return ''
      end

      # Debug helper to display applied configuration
      def debug_server_config
        debug('EC2 Server Configuration')
        debug("ec2:region '#{config[:region]}'")
        debug("ec2:availability_zone '#{config[:availability_zone]}'")
        debug("ec2:flavor_id '#{config[:flavor_id]}'")
        debug("ec2:ebs_optimized '#{config[:ebs_optimized]}'")
        debug("ec2:image_id '#{config[:image_id]}'")
        debug("ec2:security_group_ids '#{config[:security_group_ids]}'")
        debug("ec2:tags '#{config[:tags]}'")
        debug("ec2:key_name '#{config[:aws_ssh_key_id]}'")
        debug("ec2:subnet_id '#{config[:subnet_id]}'")
        debug("ec2:iam_profile_name '#{config[:iam_profile_name]}'")
        debug("ec2:username '#{config[:username]}'")
        debug("ec2:password '#{config[:password]}'")
        debug("ec2:user_data '#{config[:user_data]}'")

      end

      # Helper method for reading the Region-OS-AMI-UserName mapings to memory
      def amis
        @amis ||= begin
          json_file = File.join(File.dirname(__FILE__),
          %w{.. .. .. data amis.json})
          JSON.load(IO.read(json_file))
        end
      end

      # Networking helper
      def interface_types
        {
          'dns' => 'dns_name',
          'public' => 'public_ip_address',
          'private' => 'private_ip_address'
        }
      end

      # Helper to get the hostname for the instance

      def hostname(server)
        if config[:interface]
          method = interface_types.fetch(config[:interface]) do
            raise Kitchen::UserError, 'Invalid interface'
          end
          server.send(method)
        else
          server.dns_name || server.public_ip_address || server.private_ip_address
        end
      end

      # Get AWS Private Key
      def aws_private_key
        begin
          ENV['AWS_PRIVATE_KEY'] || ENV['AWS_SSH_KEY'] || (File.read config[:ssh_key])
        rescue
          debug('AWS_PRIVATE_KEY and AWS_SSH_KEY environment variables are not set.')
        end
      end

    end
  end
end
